src/sources/prometheus/remote_write.rs

```diff
@@ -0,0 +1,270 @@
+    fn generate_config() -> toml::Value {
+        toml::Value::try_from(Self {
+            address: "127.0.0.1:9090".parse().unwrap(),
+            tls: None,
+            auth: None,
+        })
+        .unwrap()
+    }
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "prometheus_remote_write")]
+impl SourceConfig for PrometheusRemoteWriteConfig {
+    async fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        shutdown: ShutdownSignal,
+        out: Pipeline,
+    ) -> crate::Result<sources::Source> {
+        let source = RemoteWriteSource {
+            decompressor: snap::raw::Decoder::new(),
+        };
+        source.run(self.address, "", &self.tls, &self.auth, out, shutdown)
+    }
+
+    fn output_type(&self) -> crate::config::DataType {
+        config::DataType::Metric
+    }
+
+    fn source_type(&self) -> &'static str {
+        SOURCE_NAME
+    }
+}
+
+#[derive(Clone)]
+struct RemoteWriteSource {
+    decompressor: snap::raw::Decoder,
+}
+
+impl RemoteWriteSource {
+    fn decode_body(&self, body: Bytes) -> Result<Vec<Event>, ErrorMessage> {
+        let body = self
+            .decompressor
+            .clone()
+            .decompress_vec(&body)
+            .map_err(|error| {
+                emit!(PrometheusRemoteWriteSnapError {
+                    error: error.clone()
+                });
+                ErrorMessage::new(
+                    StatusCode::BAD_REQUEST,
+                    format!("Could not decompress write request: {}", error),
+                )
+            })?;
+        let request = proto::WriteRequest::decode(Bytes::from(body)).map_err(|error| {
+            emit!(PrometheusRemoteWriteParseError {
+                error: error.clone()
+            });
+            ErrorMessage::new(
+                StatusCode::BAD_REQUEST,
+                format!("Could not decode write request: {}", error),
+            )
+        })?;
+        Ok(decode_request(request))
+    }
+}
+
+impl HttpSource for RemoteWriteSource {
+    fn build_event(
+        &self,
+        body: Bytes,
+        _header_map: HeaderMap,
+        _query_parameters: HashMap<String, String>,
+    ) -> Result<Vec<Event>, ErrorMessage> {
+        let byte_size = body.len();
+        let result = self.decode_body(body)?;
+        let count = result.len();
+        emit!(PrometheusRemoteWriteReceived { byte_size, count });
+        Ok(result)
+    }
+}
+
+fn decode_request(request: proto::WriteRequest) -> Vec<Event> {
+    request
+        .timeseries
+        .into_iter()
+        .filter_map(decode_timeseries)
+        .flatten()
+        .collect()
+}
+
+fn decode_timeseries(timeseries: proto::TimeSeries) -> Option<impl Iterator<Item = Event>> {
+    let (name, tags) = parse_labels(timeseries.labels);
+    match name {
+        Some(name) => Some(timeseries.samples.into_iter().map(move |sample| {
+            let value = sample.value;
+            let value = if name.ends_with("_total") {
```

Comment:
Do we need to document this? I could see it being a gotcha.

Reply:
Definitely would like to, I just couldn't figure out a way to fit it into the `output` section of the metadata.

Reply:
Perhaps you could add a `how_it_works` section and describe it there?